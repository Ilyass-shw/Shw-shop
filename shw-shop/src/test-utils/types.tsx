import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

export type StoreType = ReturnType<typeof configureStore>;
export type RenderReturn = ReturnType<typeof render>;
export type SliceParameters = Parameters<typeof createSlice>;
export type UiType = React.ReactElement;

export interface WrapperParams {
  MyStore: StoreType;
  WithRouter?: boolean;
}

export interface useFormProps {
  email: string;
  phone: number;
  buyerAcceptMarketing: string;
}
