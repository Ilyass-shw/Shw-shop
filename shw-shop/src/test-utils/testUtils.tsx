import React, { ReactElement } from 'react';
import { render as rtlRender, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType, useFormProps } from './types';
import { BrowserRouter } from 'react-router-dom';
import { realStoreType } from '../App/store';
import { useForm, FormProvider } from 'react-hook-form';

export const render = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(ui);
};

// ============= ============= ============= ============= =============

export const renderWithStore = (
  ui: React.ReactElement,
  MyStore: realStoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouter = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';

// ============= ============= ============= ============= =============

export const renderWithFormProvider = (
  ui: ReactElement,
  WithRouter?: 'withRouter',
  { defaultValues = {} } = {},
): RenderResult => {
  // let reactHookFormMethods = {};

  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  const WithRouterWrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return (
      <BrowserRouter>
        <FormProvider {...methods}>{children}</FormProvider>
      </BrowserRouter>
    );
  };

  if (!WithRouter) {
    return rtlRender(ui, { wrapper: Wrapper });
  }
  return rtlRender(ui, { wrapper: WithRouterWrapper });
};

// ============= ============= ============= ============= =============

export const renderWithReactHookForm = (
  ui: JSX.Element,
  saveData: jest.Mock,
): RenderResult => {
  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm<useFormProps>();
    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit((data) => {
            saveData(data);
          })}
        >
          <BrowserRouter>{children}</BrowserRouter>
          <button type="submit" data-testid="testSubmitButton"></button>
        </form>
      </FormProvider>
    );
  };
  return rtlRender(<Wrapper>{ui}</Wrapper>);
};
