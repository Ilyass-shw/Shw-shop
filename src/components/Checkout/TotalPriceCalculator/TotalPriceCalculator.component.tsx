import styled from 'styled-components';

export const PriceCalculator = styled.div`
  border-top: 1px solid #e6e6e6;
  border-bottom: 1px solid #e6e6e6;
`;

// ============= ============= ============= ============= =============

export const CalculatorBlock = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 2rem 1rem;
`;

// ============= ============= ============= ============= =============

export const Label = styled.p`
  color: #535353;
`;

// ============= ============= ============= ============= =============

export const UntilNextStep = styled.p`
  color: #535353;
  font-size: smaller;
`;

// ============= ============= ============= ============= =============

export const Price = styled.h4``;

// ============= ============= ============= ============= =============

export const Total = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 2rem 1rem;
`;

// ============= ============= ============= ============= =============

export const TotalLabel = styled.h2`
  font-size: larger;
`;

// ============= ============= ============= ============= =============

export const TotalPrice = styled.h3`
  @media screen and (min-width: 1200px) {
    font-size: x-large;
  }
`;
