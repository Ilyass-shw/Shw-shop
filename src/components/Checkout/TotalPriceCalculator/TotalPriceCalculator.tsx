import React from 'react';
import {
  PriceCalculator,
  CalculatorBlock,
  Label,
  Price,
  UntilNextStep,
  Total,
  TotalLabel,
  TotalPrice,
} from './TotalPriceCalculator.component';

interface TotalPriceCalculatorProps {
  CartTotalPrice: number;
  taxes: number;
  TotalToPay: number;
}
const TotalPriceCalculator: React.FC<TotalPriceCalculatorProps> = ({
  CartTotalPrice,
  taxes,
  TotalToPay,
}) => {
  return (
    <>
      <PriceCalculator>
        <CalculatorBlock>
          <Label>Subtotal</Label>
          <Price>{'$' + CartTotalPrice.toFixed(2)}</Price>
        </CalculatorBlock>
        <CalculatorBlock>
          <Label>Shipping</Label>
          <UntilNextStep>Calculated at next step</UntilNextStep>
        </CalculatorBlock>
        <CalculatorBlock>
          <Label>Taxes (estimated)</Label>
          <Price>{'$' + taxes.toFixed(2)}</Price>
        </CalculatorBlock>
      </PriceCalculator>
      <Total>
        <TotalLabel>Total</TotalLabel>
        <TotalPrice>{'$' + TotalToPay.toFixed(2)}</TotalPrice>
      </Total>
    </>
  );
};

export default TotalPriceCalculator;
