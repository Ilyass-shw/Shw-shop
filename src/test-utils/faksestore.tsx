import { configureStore } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { StoreType } from './types';

const fakeSlice = createSlice({
  name: 'name',
  initialState: {},
  reducers: {},
});

export default configureStore({
  reducer: {
    posts: fakeSlice.reducer,
  },
});

export const makeCustomTestStore = (
  Name: string,
  initialState: Record<string, unknown>,
  reducers = {},
): StoreType => {
  const FakeSlice = createSlice({
    name: Name,
    initialState,
    reducers,
  });

  return configureStore({
    reducer: {
      posts: FakeSlice.reducer,
    },
  });
};
