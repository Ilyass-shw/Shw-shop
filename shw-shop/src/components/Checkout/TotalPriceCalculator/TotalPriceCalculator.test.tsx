import React from 'react';
import { render } from '../../../test-utils/testUtils';
import TotalPriceCalculator from './TotalPriceCalculator';
describe('TotalPriceCalculator', () => {
  const CartTotalPrice = 56;
  const taxes = 2;
  const TotalToPay = 50;
  it('should render', () => {
    render(
      <TotalPriceCalculator
        CartTotalPrice={CartTotalPrice}
        taxes={taxes}
        TotalToPay={TotalToPay}
      />,
    );
  });
  it('should match snapshot', () => {
    const { asFragment } = render(
      <TotalPriceCalculator
        CartTotalPrice={CartTotalPrice}
        taxes={taxes}
        TotalToPay={TotalToPay}
      />,
    );
    expect(asFragment).toMatchSnapshot();
  });
});
