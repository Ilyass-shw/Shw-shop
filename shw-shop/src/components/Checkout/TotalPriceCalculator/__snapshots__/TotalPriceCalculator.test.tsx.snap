// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalPriceCalculator should match snapshot 1`] = `[Function]`;
